Compatibility-based Versioning

codello opened the pull request:
### What does this PR do?

This PR introduces a compatibility-based approach to versioning of the file format. The versioning strategy is described in the new `VERSIONING.md` file. I have decided to create one file per major version. I don't think it is necessary to split the specification further into minor versions, as minor changes will only be additions to the existing spec (and maintaining separate files would be more error-prone).

This PR also removes all "undecided" portions of the specification and removes the "work in progress" indicator from it. The intention is that after this PR the specification can be read as an official document, rather than just a current state of our discussions.

### Closes Issue(s)

Closes #57 
Closes #66 
Closes #76 
Closes #77 

### Motivation

See also #66.

The main purpose of a versioned file format is to provide certain compatibility guarantees. The introduction of a new version is usually quite the involved process, requiring cooperation of the specification team, developers, and song creators alike. Because of this, new versions should only be introduced when it is absolutely necessary.

This PR introduces the basis for a versioning strategy where a specific file format version is defined by its compatibility guarantees. As long as these guarantees can be upheld, new features are added without the need to introduce a new file format version. For transparency each version includes a revision history that outlines changes to the specification.

### More

This PR fixes #57 which is mostly a textual change without a huge impact on semantics. It is included here to avoid having two PRs, both with a very large diff.

### Additional Notes

You can use the commit history of this PR to see the changes of each file format version compared to the previous `spec.md` file.

The rest of this description includes the rationale for some of the changes:

<details>
<summary><h4>Removal of multi-valued headers</h4></summary>
Multi-valued headers were originally discussed in #22 but no decision was made. This can be discussed in a separate issue and can probably be re-introduced into the file format in a backwards-compatible way.
</details>

<details>
<summary><h4>Removal of the `CALCMEDLEY` header</h4></summary>
The current state is that the semantics of this header are not well-defined. Until we have a clear definition on the semantics, I'd consider this an unofficial header.
</details>

<details>
<summary><h4>Removal of the `NOTESGAP` header</h4></summary>
The `NOTESGAP` header was never well-defined so I have removed the current pseudo-definition.
</details>

<details>
<summary><h4>Removal of version 1.1.0</h4></summary>
Under this new versioning strategy the version 1.1.0 is essentially equivalent to version 1.0.0. I think there is no value in continuing to separate these two versions. If there is another minor release of v1 it should still be v1.2.0 to avoid confusion.
</details>

marwin89 commented:
@codello this is an outstanding contribution in ultrastar history and for me the right strategic direction. 🥇 I will review your work carefully in march.

**@rakuri255 @Baklap4 @bohning I've added you as reviewers as well, since you're important driving forces for the format. I will not merge this total change unless at least one of you consent too. But I would like to see you all consent. I would be happy if you review it in march or april.**

codello commented:
Thank you for the kind words! Please wait with your reviews until I have marked this PR as "ready for review". There are still some minor things I'd like to change (and need to resolve the merge conflict). The PR should be ready in a few days.

codello committed:
Split spec.md

Going forward the file format specification is maintained in separate files for each major version.

codello committed:
Rewrite unversioned file format

The unversioned file format describes the widely-used UltraStar file format without a #VERSION header.

codello committed:
Rewrite UltraStar file format v1

codello committed:
Rewrite UltraStar file format v2

codello committed:
Add VERSIONING.md

codello commented:
The PR is ready for review now.

One thing that is not part of this PR is the publication process for new versions (or revisions). This has been touched on in #66. Specifically the following situations are not addressed by the versioning strategy in this PR:
- Introduction of a change that should be reverted shortly afterwards
- Aggregation of multiple (editorial) changes to keep the revision history short
- Planning of publications for a certain point in time.

I think it could be interesting to adopt a **branch-based approach**. Development of the specification could be done in a _development_ branch. Multiple changes can be aggregated (and potentially reverted) there. The changes are not considered "published" yet.
When we want to publish a new revision of a version we can merge the development branch into main, thereby making the changes official. If we want to have different publication schedules for different versions of the spec we could use more than one dev branch.

This is probably better discussed in a separate issue. I wanted to include the suggestion here because the versioning strategy by itself doesn't concern itself with a publication strategy for new revisions of the spec, although it might be of interest (see https://github.com/UltraStar-Deluxe/format/issues/66#issuecomment-2612473506).

TheNailDev commented:
I like the direction of the new compatibility based versioning. Thank you for your work. :+1: 

Looking at the changes I noticed one thing regarding the note section.
The original spec contained the following note regarding unknown note types (`spec.md`, line 795):
```md
Implementations MAY substitute unknown note types with freestyle notes (`F`).
```
This is not included in the new specification, which is fine IMO. But I think it should be explicitly specified that implementations should not (or even must not?) reject a song based on it containing unknown note types. Otherwise the example for a non-breaking change impacting compatibility would actually be a breaking change (`versioning.md`, lines 102-105):
```md
**Example**: Adding a new note type (let's say a vibrato note).
A song that uses this note type can be opened in a game that does not support it and can be sung, but the vibrato notes will not work.
They will be either missing or will be replaced by Freestyle notes.
The song can still be sung, but players will experience some restrictions when newer features are used.
```

codello commented:
Ah yes, great catch with the `versioning.md`. I'm gonna try to find a better example there.

I actually intentionally introduced this ambiguity in the PR. In theory this PR should not change any of the contents/semantics of the specification but just its presentation (multiple files vs 1 file and a different structure within the file). Previously there was the following warning indicating that handling of unknown note isn't really well-defined yet:

https://github.com/UltraStar-Deluxe/format/blob/5d39a2b98be0912821c0c05a4c972a8c55d645be/spec.md?plain=1#L803-L805

The whole "custom note types" is being discussed in #72 and I didn't want to anticipate a result of that discussion in this PR (seems like I did anyways with the example in `versioning.md`). I felt that it was closest to the current state of the discussion to just leave the handling of unknown note types undefined, so that we can define it via #72. I am personally hoping that we can just reintroduce the previous phrasing, allowing implementations to ignore note types that they don't recognize.

marwin89 reviewed (commented):
@codello , the 3 specifications and the versioning description look pretty flawless to me.

I will approve this PR next monday. @Baklap4 @bohning is there any blocker for you?
